src/main/java/org/littleshoot/proxy/impl/ProxyToServerConnection.java

```diff
@@ -541,51 +541,75 @@ private void connectAndWrite(HttpRequest initialRequest) {
+        if (ProxyUtils.isCONNECT(initialRequest) && isMitmEnabled() &&
+                (!currentFilters.shouldInitiateServerSSLHandshake() || remoteAddress.isUnresolved())) {  // Skip SSL handshake
+            connectionFlow.then(clientConnection.RespondCONNECTSuccessful);
+            connectionFlow.then(serverConnection.MitmEncryptClientChannel);
+        } else {
+            connectionFlow.then(ConnectChannel);
+
+            if (chainedProxy != null && chainedProxy.requiresEncryption()) {
+                connectionFlow.then(serverConnection.EncryptChannel(chainedProxy
+                        .newSslEngine()));
+            }
+
+            if (ProxyUtils.isCONNECT(initialRequest)) {
+                // If we're chaining, forward the CONNECT request
+                if (hasUpstreamChainedProxy()) {
+                    connectionFlow.then(
+                            serverConnection.HTTPCONNECTWithChainedProxy);
                 }
 
-            	connectionFlow
-                        .then(clientConnection.RespondCONNECTSuccessful)
-                        .then(serverConnection.MitmEncryptClientChannel);
-            } else {
-                connectionFlow.then(serverConnection.StartTunneling)
-                        .then(clientConnection.RespondCONNECTSuccessful)
-                        .then(clientConnection.StartTunneling);
+
+                if (isMitmEnabled()) {
+                    // When MITM is enabled and when chained proxy is set up, remoteAddress
+                    // will be the chained proxy's address. So we use serverHostAndPort
+                    // which is the end server's address.
+                    HostAndPort parsedHostAndPort = HostAndPort.fromString(serverHostAndPort);
+
+                    // SNI may be disabled for this request due to a previous failed attempt to connect to the server
+                    // with SNI enabled.
+                    if (disableSni) {
+                        connectionFlow.then(serverConnection.EncryptChannel(proxyServer.getMitmManager()
+                                .serverSslEngine()));
+                    } else {
+                        connectionFlow.then(serverConnection.EncryptChannel(proxyServer.getMitmManager()
+                                .serverSslEngine(parsedHostAndPort.getHost(), parsedHostAndPort.getPort())));
+                    }
+
+                    connectionFlow
+                            .then(clientConnection.RespondCONNECTSuccessful)
+                            .then(serverConnection.MitmEncryptClientChannel);
+                } else {
+                    connectionFlow.then(serverConnection.StartTunneling)
+                            .then(clientConnection.RespondCONNECTSuccessful)
+                            .then(clientConnection.StartTunneling);
+                }
             }
         }
     }
 
+    /**
+     *
+     * @return SSLSession or Null
+     */
+    private SSLSession getSSLSessionsOrNull() {
```

Comment:
I think this could just be: ```private @Nullable SSLSession getSSLSession() { ... }``` which will double for IDE producing warnings wherever this method gets called without checking for null

Reply:
+1